src/libraries/System.Private.CoreLib/src/System/Random.cs

```diff
@@ -1,197 +1,599 @@
+
+            public override void NextBytes(byte[] buffer) => NextBytes((Span<byte>)buffer);
+
+            public override void NextBytes(Span<byte> buffer)
+            {
+                ulong next;
+
+                while (buffer.Length >= sizeof(long))
+                {
+                    next = NextUInt64();
+                    MemoryMarshal.Write(buffer, ref next);
+                    buffer = buffer.Slice(sizeof(long));
+                }
+
+                next = NextUInt64();
+                unsafe
+                {
+                    byte* remainingBytes = (byte*)&next;
+                    for (int i = 0; i < buffer.Length; i++)
+                    {
+                        buffer[i] = remainingBytes[i];
+                    }
+                }
+            }
+
+            public override double NextDouble() =>
+                // As described in http://prng.di.unimi.it/:
+                // "A standard double (64-bit) floating-point number in IEEE floating point format has 52 bits of significand,
+                //  plus an implicit bit at the left of the significand. Thus, the representation can actually store numbers with
+                //  53 significant binary digits. Because of this fact, in C99 a 64-bit unsigned integer x should be converted to
+                //  a 64-bit double using the expression
+                //  (x >> 11) * 0x1.0p-53"
+                (NextUInt64() >> 11) * (1.0 / (1ul << 53));
+
+            public override float NextSingle() =>
+                // Same as above, but with 24 bits instead of 53.
+                (NextUInt64() >> 40) * (1.0f / (1u << 24));
+
+            public override double Sample()
+            {
+                Debug.Fail("Not used or called for this implementation.");
+                throw new NotSupportedException();
+            }
         }
 
-        private static Random ThreadStaticRandom
+        /// <summary>
+        /// Provides an implementation used for compatibility with cases where either a) the
+        /// sequence of numbers could be predicted based on the algorithm employed historically and
+        /// thus expected (e.g. a specific seed used in tests) or b) where a derived type may
+        /// reasonably expect its overrides to be called.  The algorithm is based on a modified version
+        /// of Knuth's subtractive random number generator algorithm.  See https://github.com/dotnet/runtime/issues/23198
+        /// for a discussion of some of the modifications / discrepancies.
+        /// </summary>
+        private sealed class LegacyImpl : ImplBase
         {
-            get
+            [ThreadStatic]
+            private static Xoshiro256StarStarImpl? t_seedGenerator;
+
+            private readonly Random _parent;
+            private readonly int[] _seedArray;
+            private int _inext;
+            private int _inextp;
+
+            public LegacyImpl(Random parent) : this(parent, (t_seedGenerator ??= new()).Next())
```

Comment:
Would you mind describing why the parent's instance is being passed to the `LegacyImpl` constructor? I see that `_parent` is being used to call its `Sample` method, but I'm failing to understand why the legacy code was not moved to `LegacyImpl` to have the `Random` APIs call them directly.

Reply:
Most of the methods, e.g. Sample, on Random are virtual.  And some of the methods delegate to other methods, e.g. Next(int, int) calls Sample.  So if the derived type has overridden Sample, it's expecting that the base implementation of Next(int, int) will call its override.